src/genesis/mod.rs

```diff
@@ -139,42 +142,60 @@ where
 /// header.
 async fn load_and_retrieve_header<DB, R>(
     store: DB,
-    mut reader: R,
+    reader: R,
     skip_load: bool,
 ) -> anyhow::Result<(Vec<Cid>, Option<usize>)>
 where
     DB: Blockstore + Send + Sync + 'static,
     R: AsyncRead + Send + Unpin,
 {
     let result = if skip_load {
-        (CarReader::new(&mut reader).await?.header.roots, None)
+        (CarReader::new(reader).await?.header.roots, None)
     } else {
-        let (roots, n_records) = forest_load_car(store, &mut reader).await?;
+        let (roots, n_records) = forest_load_car(store, reader).await?;
         (roots, Some(n_records))
     };
 
     Ok(result)
 }
 
+fn car_stream<R: futures::AsyncRead + Send + Unpin>(
+    mut reader: CarReader<R>,
+) -> impl Stream<Item = anyhow::Result<fvm_ipld_car::Block>> {
+    try_stream! {
+        while let Some(block) = reader.next_block().await? {
+            yield block;
+        }
+    }
+}
+
 pub async fn forest_load_car<DB, R>(store: DB, reader: R) -> anyhow::Result<(Vec<Cid>, usize)>
 where
     R: futures::AsyncRead + Send + Unpin,
     DB: Blockstore + Send + Sync + 'static,
 {
-    // 1GB
-    const BUFFER_CAPCITY_BYTES: usize = 1024 * 1024 * 1024;
+    let db = Arc::new(store);
 
-    let (tx, rx) = flume::bounded(100);
-    let write_task =
-        tokio::spawn(async move { store.buffered_write(rx, BUFFER_CAPCITY_BYTES).await });
-    let mut car_reader = CarReader::new(reader).await?;
+    let car_reader = CarReader::new(reader).await?;
+    let roots = car_reader.header.roots.clone();
     let mut n_records = 0;
-    while let Some(block) = car_reader.next_block().await? {
-        debug!("Importing block: {}", block.cid);
-        n_records += 1;
-        tx.send_async((block.cid, block.data)).await?;
-    }
-    drop(tx);
-    write_task.await??;
-    Ok((car_reader.header.roots, n_records))
+
+    let sink = |block: Vec<fvm_ipld_car::Block>| {
+        let db = Arc::clone(&db);
+        async {
+            tokio::task::spawn_blocking(move || {
```

Comment:
And this here I assume is done for back-pressure?

Reply:
Committing to ParityDB can potentially block indefinitely.